[Inductor] Node Level provenance tracking

yushangdi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142739

- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.


Added a hook to graph module deepcopy, because if we deepcopied a gm within the context, the hooks are also copied over, and we need to remove the hook from the new deepcopied gm. 

For example, if we have 
```
        with GraphTransformObserver(gm, "test") as obs:
            gm2 = copy.deepcopy(gm)
```

We want to make sure that `gm2._replace_hooks`(and other hooks) is empty. If we do not explicitly remove the hooks on gm2, gm2 will contains the hooks copied from gm. Because gm has those hooks within the context, and `obs` only removes the hook from `gm` when exiting the context.

https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

yushangdi committed:
[Inductor] Node Level provenance tracking

- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.



https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

[ghstack-poisoned]

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.



https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.



https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/compile_fx.py:
```diff
@@ -918,11 +918,17 @@ def log_graph_runnable() -> str:
-                    ),
+                    )
+                    + "# "
+                    + provenance_tracking_json,
```
I am not sure about the answer, but will changing the log here confuse tlparse?

yushangdi replied:
I've tried it on tlparse, and tlparse will just have this additional comment line in the resulting post_grad_graph graph file. 

> I am not sure about the answer, but will changing the log here confuse tlparse?

yushangdi replied:
@desertfire some update on this line: it's not confusing tlparse, but it does break some tests that test the content of `inductor_post_grad_graph`. So I'm removing this line. I'll make another PR that dump this json separately.

desertfire commented on torch/fx/passes/graph_transform_observer.py:
```diff
@@ -83,22 +84,41 @@ def _check_disable_pass(self):
         self.gm._unregister_erase_node_hook(self.on_node_erase)
+        self.gm._unregister_replace_node_hook(self.on_node_replace)
+
+        for node in self.gm.graph.nodes:
```
Why do we need the following explicit propagation here? Does it mean "node.meta" is still not properly maintained at some places?

yushangdi replied:
> Why do we need the following explicit propagation here? Does it mean "node.meta" is still not properly maintained at some places?

In some sense, yes. The node meta is simply copied over from the old node at the time of node creation or node replacement. So here we update the node meta. 

An alternative is we change the create_node_hook and replace_node_hook so node meta is updated at the time of node creation/node replacement.

desertfire replied:
I would prefer doing the propagation eagerly than lazily, in case anything in the middle of a pass needs the most up-to-date information.

yushangdi commented:
> @Chillee , should we move https://github.com/Chillee/compiler_explorer to core repo?

FYI, @desertfire @Chillee I made a fork of Horace's repo here: https://github.com/yushangdi/compiler_explorer. Both me and Rachel are working on this forked repo now.

Chillee commented:
@yushangdi I don't mind - I think it would make sense to move it eventually.

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.



https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.



https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.


Added a hook to graph module deepcopy, because if we deepcopied a gm within the context, the hooks are also copied over, and we need to remove the hook from the new deepcopied gm. 

For example, if we have 
```
        with GraphTransformObserver(gm, "test") as obs:
            gm2 = copy.deepcopy(gm)
```

We want to make sure that `gm2._replace_hooks`(and other hooks) is empty. If we do not explicitly remove the hooks on gm2, gm2 will contains the hooks copied from gm. Because gm has those hooks within the context, and `obs` only removes the hook from `gm` when exiting the context.

https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yushangdi committed:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.


Added a hook to graph module deepcopy, because if we deepcopied a gm within the context, the hooks are also copied over, and we need to remove the hook from the new deepcopied gm. 

For example, if we have 
```
        with GraphTransformObserver(gm, "test") as obs:
            gm2 = copy.deepcopy(gm)
```

We want to make sure that `gm2._replace_hooks`(and other hooks) is empty. If we do not explicitly remove the hooks on gm2, gm2 will contains the hooks copied from gm. Because gm has those hooks within the context, and `obs` only removes the hook from `gm` when exiting the context.

https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

desertfire reviewed (approved):
LGTM. cc @bdhirsh to see if he has more comments on the fx part

Commit:
Update on "[Inductor] Node Level provenance tracking"


- use GraphTransformObserver + replace_node hooks to track node sources when they are replaced
- add pre_grad_graph tracking to tlparse
- add the node provenance information to post_grad_graph tlparse. This is for the frontend to create a mapping between pre_grad and post_grad graph. See an example frontend (this is just a prototype) here:  https://drive.google.com/file/d/1cMHH_0y4FJUSS9tATwGQvA72O0Lth8eh/view?usp=sharing
- change "action" of NodeSource from a single action to a list of actions.


Added a hook to graph module deepcopy, because if we deepcopied a gm within the context, the hooks are also copied over, and we need to remove the hook from the new deepcopied gm. 

For example, if we have 
```
        with GraphTransformObserver(gm, "test") as obs:
            gm2 = copy.deepcopy(gm)
```

We want to make sure that `gm2._replace_hooks`(and other hooks) is empty. If we do not explicitly remove the hooks on gm2, gm2 will contains the hooks copied from gm. Because gm has those hooks within the context, and `obs` only removes the hook from `gm` when exiting the context.

https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?tab=t.0

The front-end code that takes in the tlparse result is in https://github.com/yushangdi/compiler_explorer.

Differential Revision: [D65006709](https://our.internmc.facebook.com/intern/diff/D65006709/)

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yushangdi commented:
close. The updated PR in  https://github.com/pytorch/pytorch/pull/144277
